### ISSUE TITLE
Rule proposal: `prefer-immediate-return`

### ISSUE THREAD

**Issue by remcohaszing** [sindresorhus/eslint-plugin-unicorn#1520]
In my opinion it’s never desirable to assign a value to a variable, only to be returned immediately. I believe this is often the leftover result of debugging.

In other words, if a value is declared, and returned immediately in the next statement, the two should be merged, removing the useless variable.

## Fail

```js
// A
function foo() {
  const result = bar();
  return result;
}
```

```js
// B
function foo() {
  let result;
  return result;
}
```

## Pass

```js
function foo() {
  const result = bar();
  // Anything in between is allowed.
  console.log(result);
  return result;
}
```

```js
function foo() {
  let result;
  result = bar();
  return result;
}
```

```js
// Autofix result for A
function foo() {
  return bar();
}
```

```js
// Autofix result for B
function foo() {
  return;
}
```

**Comment by fisker** [comment]
There is an existing rule [`sonarjs/prefer-immediate-return`](https://github.com/SonarSource/eslint-plugin-sonarjs/blob/master/docs/rules/prefer-immediate-return.md), but I'm not a fan, I [disabled it recently](https://github.com/fisker/shared-configs/commit/0f4aaf63580f286304bdc2aeed0e9f281e479b83).

**Comment by remcohaszing** [comment]
You’re right, it even has the same name as I came up with. I can’t believe I didn’t find this earlier. :sweat_smile: 

I don’t know if `eslint-plugin-unicorn` wants to duplicate rules from other plugins, so I’ll just leave this issue open for now.

**Comment by fregante** [comment]
@fisker Why not a fan? It seems reasonable. When would that code be preferable?

**Comment by fisker** [comment]
I forgot where the code is, but this is similar to "anonymous default export", sometimes I want to give a name to the return value.

**Comment by slikts** [comment]
This rule conflicts with returning named functions, which is useful for debuggability:
```js
const makeFoo = () => {
  const foo = () => {
    // something
  }
  return foo
}
```

**Comment by fregante** [comment]
> I forgot where the code is, but this is similar to "anonymous default export"

You can use use `return function foo() {}` just like with `export default function foo() {}`

> sometimes I want to give a name to the return value.

What's the purpose of this? The name of non-functions variables is lost after return (i.e. it's not available outside the function), so this would only be useful on those exact 2 lines:

```js
const life = 42;
return life;
``` 


> returning named functions, which is useful for debuggability

I think that's already covered (i.e. forbidden) by `func-names`, example:

```js
export function a() {
	return function foo() {
		return 1
	}
}
```
```
✖  220:9   Unexpected named function foo.               func-names
```

If you want to name functions, you can disable `func-names` instead (which has been enabled for years in XO, for example)

---

I think this rule would conflict exclusively if you intend to use named arrow functions, which… seems picky since regular functions can be named _and_ can be returned immediately.

**Comment by remcohaszing** [comment]
This rule would also play nice with [`no-return-await`](https://eslint.org/docs/rules/no-return-await). Currently an intermediate variable could be used to circumvent this rule.

```js
async function foo() {
  const value = await bar();
  return value;
}
```

Becomes (because of `prefer-immediate-return`):

```js
async function foo() {
  return await bar();
}
```

Which then becomes (because of `no-return-await`):

```js
async function foo() {
  return bar();
}
```

**Comment by slikts** [comment]
@fregante I'm using arrow functions for consistency and clarity and don't want to have to switch to named function expressions. I've made a request to support this use case at SonarSource/eslint-plugin-sonarjs#293.

**Comment by remcohaszing** [comment]
There could be an option `allowArrowFunctions`.

**Comment by fregante** [comment]
> I'm using arrow functions for consistency and clarity

As I said, that seems picky. Nothing clear about 

```js
const makeFoo = () => {
  const foo = () => {
    // something
  }
  return foo
}
```
over

```js
const makeFoo = () => {
  return function foo() {
    // something
  }
}
```

Or even

```js
function makeFoo() {
  return function foo() {
    // something
  }
}
```

If anything, it's the opposite: It's more compact and has less `=` `>` `(` `)` noise.



> since it means the `.name` property for functions is different.

If you're using a minifier those names are being mangled to begin with.

**Comment by slikts** [comment]
Using both arrow functions and function expressions is inconsistent, because, if primarily using arrow functions, immediately returning a named function would be an exception. Function declarations also have undesirable behaviors like `foo(); function foo() {}` being possible. Arrow functions also communicate clearly that `this` is not used, and arrow function syntax is more terse. Spec compliant transpilers will support function name inference, although it's for debugging anyway.

**Comment by fregante** [comment]
> arrow function syntax is more terse

Not

```js
const foo = () => {
};
function foo() {
}
``` 

> Function declarations also have undesirable behaviors like foo(); function foo() {} being possible

I don’t see a problem with hoisting

**Comment by slikts** [comment]
This is an example of the issues with hoisting, in that they cause confusion. `const` is also hoisted, just not initialized. You can do `const foo = () => bar(); const bar = () => {}; foo()` just fine. With function declarations, however, the function definition is also hoisted (not just the declaration), making them more of a special case and so making their hoisting behavior more confusing for no benefit.

Function declarations are also more loose compared to `const` because you can re-declare them, so it's an another source of mistakes.

The arrow function syntax is more terse for anonymous functions and/or when `return` is omitted. 

By "not using `this`" I obviously mean that `this` can only be bound lexically in arrow functions and can't be rebound dynamically. Using arrow functions makes the intentions clear, whereas sometimes using functions expressions instead of arrow would require anyone reading the code to consider if the context isn't rebound.

In any case, the use case I mentioned (consistent use of arrow functions and returning named functions) is valid and your stylistic preferences of mixing types of function definitions are not really relevant.

**Comment by fregante** [comment]
I updated my comment 👆

**Comment by slikts** [comment]
I saw the updates before replying, and the last line addresses them: your personal stylistic preferences don't make my use case (consistently using arrow functions) invalid.

**Comment by fregante** [comment]
You're  wrong and confused. You're the one arguing from a stylistic standpoint.

> consistent use of arrow functions and returning named functions

> arrow function syntax is more terse

These are stylistic preferences and you said them.






> Function declarations are also more loose compared to `const` because you can re-declare them, so it's an another source of mistakes.

Invalid. `return function foo() {}` is not a declaration.



> This is an example of the issues with hoisting, in that they cause confusion

Invalid. There's no hoisting in `return function foo () {}`, try it.



> The arrow function syntax is more terse for anonymous functions

Invalid. **You're here to ask for functions to be named so they're *not** anonymous.**

**Comment by slikts** [comment]
I'd suggest trying to practice charitable interpretation, since I'm referencing arrow functions and function declarations and expressions separately and in the correct contexts, so I know the difference. Declarations are relevant as part of the general reason to prefer consistent function definitions using only the arrow syntax, since declarations are one of the two alternatives (ignoring the Function constructor). That's a preference with valid reasons and you not seeing an issue with hoisting, etc. doesn't override those reasons, and function expressions are not relevant since that'd mean giving up consistent function definitions.

**Comment by AndreaPontrandolfo** [comment]
Agree with @fisker , it's one of the few sonarjs rules that i always leave off. 
It's one of those rules that seems good in theory but then do more harm then good in practice.
Having that variable helps with readability, discovery/searchability and debugging.

**Comment by sindresorhus** [comment]
Going to pass on this.

The rule is mostly stylistic, and the cases where the temporary variable is useful are common enough: naming an intermediate result for readability, debugging, stack traces, or searchability.
